File: src/gtm_sourcing_agent/task_queue.py
"""In-process background task worker — Phase 4 (async + scale,
docs/product-plan.md). Every LLM-touching stage call in api.py is
enqueued here instead of running inline in the request handler, so a
slow real model call (real Anthropic latency, not the mock's instant
canned response) never blocks the HTTP response or ties up a request
thread for its whole duration.

Deliberately a single dedicated worker thread pulling task ids off a
queue.Queue, not a pool. Concurrent writer threads against the SQLite
file caused real "database is locked" / "attempt to write a readonly
database" failures during development (see db.py) — a single sequential
worker sidesteps that entirely, with no WAL-mode tuning and no external
broker (Celery/Redis) this product doesn't need yet at this scale. If
job volume ever outgrows one worker, swap the queue for a real broker;
nothing above task_queue.py (api.py's routes) would need to change,
same "swap storage backends, never change what's above them" pattern
Phase 1 established for db_storage.py.

This module knows nothing about stages/*.py — api.py registers a
zero-context runner per task `kind` via register_runner(), keeping the
dependency direction one-way (api.py -> stages/*.py), unchanged from
before this phase.
"""

import logging
import queue
import threading
from typing import Any, Callable

from . import db_storage

logger = logging.getLogger(__name__)

TaskRunner = Callable[[str, dict[str, Any]], dict[str, Any]]

_queue: "queue.Queue[str]" = queue.Queue()
_runners: dict[str, TaskRunner] = {}
_worker_started = False
_start_lock = threading.Lock()


def _recover_orphaned_tasks() -> None:
    """Runs once per process, from _ensure_worker() below, before the
    worker thread starts and before this call's own enqueue() creates
    its task row. This process's in-memory queue starts empty regardless
    of what the database says — so any task left "pending" or "running"
    from a previous process (crash, redeploy) is orphaned: nothing will
    ever finish it. Mark those failed so the recruiter sees an honest
    "this didn't complete, try again" instead of a spinner that never
    resolves.

    This assumes exactly one server process/instance, same as the rest
    of this module (see the module docstring). Running more than one
    instance would make this actively wrong — a second instance's
    startup would mark the first instance's genuinely in-flight tasks
    as failed. Do not scale this app horizontally without replacing the
    in-memory queue with something that coordinates across processes.
    """
    count = db_storage.reset_incomplete_tasks(
        "Interrupted by a server restart before it finished. Please retry."
    )
    if count:
        logger.warning("task worker: recovered %d orphaned task(s) left by a previous process", count)


def register_runner(kind: str, fn: TaskRunner) -> None:
    """Register how to execute a task of the given `kind`. `fn(role_id,
    args)` must return a JSON-serializable result (a stage's own
    `.model_dump()`) or raise ValueError/RuntimeError, same contract
    stage functions already have via storage_backend."""
    _runners[kind] = fn


def enqueue(role_id: str, kind: str, args: dict[str, Any]) -> dict[str, Any]:
    if kind not in _runners:
        raise ValueError(f"no task runner registered for kind '{kind}'")
    # Recovery must run before this task's own row is created below —
    # otherwise the very first enqueue() of a process would immediately
    # mark its own brand-new "pending" task as orphaned.
    _ensure_worker()
    task = db_storage.create_task(role_id, kind, args)
    _queue.put(task["task_id"])
    return task


def _ensure_worker() -> None:
    global _worker_started
    with _start_lock:
        if _worker_started:
            return
        _recover_orphaned_tasks()
        threading.Thread(target=_worker_loop, name="gtm-task-worker", daemon=True).start()
        _worker_started = True


def _worker_loop() -> None:
    while True:
        task_id = _queue.get()
        try:
            _run_one(task_id)
        except Exception:  # the worker thread must never die — one bad task shouldn't wedge every task after it
            logger.exception("task worker: unhandled error processing task %s", task_id)
        finally:
            _queue.task_done()


def _run_one(task_id: str) -> None:
    task = db_storage.get_task(task_id)
    if task is None:
        logger.warning("task worker: task %s vanished before it could run", task_id)
        return
    runner = _runners.get(task["kind"])
    if runner is None:
        db_storage.update_task(task_id, status="failed", error=f"no runner registered for kind '{task['kind']}'")
        return
    db_storage.update_task(task_id, status="running")
    try:
        result = runner(task["role_id"], task["args"])
    except (ValueError, RuntimeError) as e:
        # same two exception types _run_stage used to map to 400/502 — now
        # surfaced as a failed task's .error instead of an HTTP status,
        # since there's no request left open to carry a status code by the
        # time a real model call finishes.
        db_storage.update_task(task_id, status="failed", error=str(e))
    except Exception as e:  # unexpected — still surface it, never leave a task stuck "running" forever
        logger.exception("task worker: unexpected error running task %s (kind=%s)", task_id, task["kind"])
        db_storage.update_task(task_id, status="failed", error=f"unexpected error: {e}")
    else:
        db_storage.update_task(task_id, status="succeeded", result=result)

File: backend-node/src/taskQueue.ts
/**
 * Port of task_queue.py -- in-process background task worker. Every
 * LLM-touching stage call is enqueued here instead of running inline in
 * a route handler, so a slow real model call never blocks the HTTP
 * response.
 *
 * Deliberately a single sequential in-memory queue, not a pool --
 * carries over the exact same "single server process" constraint
 * Python's module docstring documents (its reason was SQLite lock
 * contention across worker *threads*; Node has no threads to contend in
 * the first place, but the queue is still in-process memory, so running
 * more than one Node process/instance would still leave each instance
 * with its own isolated queue and its own view of "what's running" --
 * the same horizontal-scaling limitation, not fixed, not hidden, exactly
 * per the migration brief's instruction to carry over what the Python
 * code honestly documents rather than silently fixing it. If job volume
 * ever outgrows one process, swap this module for a real broker; nothing
 * above it (route handlers) would need to change.
 */
import * as storage from "./db/storage.js";
import { StorageError } from "./db/storage.js";
import { LlmError } from "./llmClient.js";

export type TaskRunner = (roleId: string, args: Record<string, any>) => Promise<Record<string, any>>;

const queue: string[] = [];
const runners = new Map<string, TaskRunner>();
let workerStarted = false;
let draining: Promise<void> | null = null;

export function registerRunner(kind: string, fn: TaskRunner): void {
  runners.set(kind, fn);
}

export async function enqueue(roleId: string, kind: string, args: Record<string, any>) {
  if (!runners.has(kind)) {
    throw new StorageError(`no task runner registered for kind '${kind}'`);
  }
  // Recovery must run before this task's own row is created below --
  // otherwise the very first enqueue() of a process would immediately
  // mark its own brand-new "pending" task as orphaned.
  await ensureWorker();
  const task = await storage.createTask(roleId, kind, args);
  queue.push(task.task_id);
  kickWorker();
  return task;
}

async function ensureWorker(): Promise<void> {
  if (workerStarted) return;
  workerStarted = true;
  await recoverOrphanedTasks();
}

async function recoverOrphanedTasks(): Promise<void> {
  // Runs once per process, before this process's own in-memory queue
  // (which always starts empty) has anything in it -- any task still
  // "pending"/"running" in the DB from a previous process (crash,
  // redeploy) is orphaned: nothing will ever finish it.
  const count = await storage.resetIncompleteTasks("Interrupted by a server restart before it finished. Please retry.");
  if (count) {
    console.warn(`task worker: recovered ${count} orphaned task(s) left by a previous process`);
  }
}

function kickWorker(): void {
  if (draining) return; // already processing the queue
  draining = drain().finally(() => {
    draining = null;
  });
}

async function drain(): Promise<void> {
  let taskId: string | undefined;
  while ((taskId = queue.shift()) !== undefined) {
    try {
      await runOne(taskId);
    } catch (e) {
      // the worker loop must never die -- one bad task shouldn't wedge every task after it
      console.error(`task worker: unhandled error processing task ${taskId}`, e);
    }
  }
}

async function runOne(taskId: string): Promise<void> {
  const task = await storage.getTask(taskId);
  if (task === null) {
    console.warn(`task worker: task ${taskId} vanished before it could run`);
    return;
  }
  const runner = runners.get(task.kind);
  if (!runner) {
    await storage.updateTask(taskId, { status: "failed", error: `no runner registered for kind '${task.kind}'` });
    return;
  }
  await storage.updateTask(taskId, { status: "running" });
  try {
    const result = await runner(task.role_id, task.args as Record<string, any>);
    await storage.updateTask(taskId, { status: "succeeded", result });
  } catch (e: any) {
    // StorageError/LlmError are the same two categories _runStage maps
    // to 400/502 on a synchronous route -- surfaced here as a failed
    // task's .error instead, since there's no request left open to carry
    // a status code by the time a real model call finishes.
    if (e instanceof StorageError || e instanceof LlmError) {
      await storage.updateTask(taskId, { status: "failed", error: e.message });
      return;
    }
    console.error(`task worker: unexpected error running task ${taskId} (kind=${task.kind})`, e);
    await storage.updateTask(taskId, { status: "failed", error: `unexpected error: ${e?.message ?? e}` });
  }
}
